packages/xstate-react/src/useInterpret.ts

```diff
@@ -0,0 +1,110 @@
+import { useState, useEffect } from 'react';
+import useIsomorphicLayoutEffect from 'use-isomorphic-layout-effect';
+import {
+  interpret,
+  EventObject,
+  StateMachine,
+  State,
+  Interpreter,
+  InterpreterOptions,
+  MachineOptions,
+  Typestate,
+  Observer
+} from 'xstate';
+import { toObserver } from 'xstate/lib/utils';
+import { MaybeLazy } from './types';
+import useConstant from './useConstant';
+import { UseMachineOptions } from './useMachine';
+import { useReactEffectActions } from './useReactEffectActions';
+
+export function useInterpret<
+  TContext,
+  TEvent extends EventObject,
+  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+>(
+  getMachine: MaybeLazy<StateMachine<TContext, any, TEvent, TTypestate>>,
+  options: Partial<InterpreterOptions> &
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>> = {},
+  observerOrListener?:
+    | Observer<State<TContext, TEvent, any, TTypestate>>
+    | ((value: State<TContext, TEvent, any, TTypestate>) => void)
+): Interpreter<TContext, any, TEvent, TTypestate> {
+  const machine = useConstant(() => {
+    return typeof getMachine === 'function' ? getMachine() : getMachine;
+  });
+
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    typeof getMachine !== 'function'
+  ) {
+    const [initialMachine] = useState(machine);
+
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+    const machineWithConfig = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    return interpret(machineWithConfig, {
+      deferEvents: true,
+      ...interpreterOptions
+    });
+  });
+
+  useIsomorphicLayoutEffect(() => {
+    if (observerOrListener) {
+      service.subscribe(toObserver(observerOrListener));
```

Comment:
This subscription will suffer from a stale closure problem - we probably should avoid that. Maybe the easiest way to handle this would be to:
1. only allow function listener right now - they are the most useful anyway
2. use a pattern like this:
```js
const listenerRef = React.useRef(listener)
useIsomorphicLayoutEffect(() => (listenerRef.current = listener))

useIsomorphicLayoutEffect(() => {
  service.subscribe(state => {
    const currentListener = listenerRef.current
    currentListener && currentListener(state)
  })
  // ...
})
```

The first part of this React dance (😰) is what I've encapsulated in the [`useLatest`](https://github.com/Andarist/use-latest) helper as I had already handled this just too many times.

There is also a quite good, recent, read touching the callbacks+refs subject: https://dev.to/yelouafi/a-simple-rule-for-using-callbacks-in-react-4jah

Reply:
Ref might not be needed:

```jsx
  useIsomorphicLayoutEffect(() => {
    let sub;
    if (observerOrListener) {
      sub = service.subscribe(toObserver(observerOrListener));
    }

    return () => {
      sub?.unsubscribe();
    };
  }, [observerOrListener]);
```